components/places/src/storage/history.rs

```diff
@@ -1037,15 +1037,67 @@ pub fn get_visit_infos(
     db: &PlacesDb,
     start: Timestamp,
     end: Timestamp,
+    exclude_types: VisitTransitionSet,
 ) -> Result<HistoryVisitInfos> {
+    let allowed_types = exclude_types.complement();
     let infos = db.query_rows_and_then_named_cached(
         "SELECT h.url, h.title, v.visit_date, v.visit_type
          FROM moz_places h
          JOIN moz_historyvisits v
            ON h.id = v.place_id
          WHERE v.visit_date BETWEEN :start AND :end
+           AND ((1 << visit_type) & :allowed_types) != 0
          ORDER BY v.visit_date",
-        &[(":start", &start), (":end", &end)],
+        rusqlite::named_params! {
+            ":start": start,
+            ":end": end,
+            ":allowed_types": allowed_types,
+        },
+        HistoryVisitInfo::from_row,
+    )?;
+    Ok(HistoryVisitInfos { infos })
+}
+
+pub fn get_visit_count(db: &PlacesDb, exclude_types: VisitTransitionSet) -> Result<i64> {
+    let count = if exclude_types.is_empty() {
+        db.query_one::<i64>("SELECT COUNT(*) FROM moz_historyvisits")?
+    } else {
+        let allowed_types = exclude_types.complement();
+        db.query_row_and_then_named(
+            "SELECT COUNT(*)
+             FROM moz_historyvisits
+             WHERE ((1 << visit_type) & :allowed_types) != 0",
+            rusqlite::named_params! {
+                ":allowed_types": allowed_types,
+            },
+            |r| r.get(0),
+            true,
+        )?
+    };
+    Ok(count)
+}
+
+pub fn get_visit_page(
+    db: &PlacesDb,
+    offset: i64,
+    count: i64,
+    exclude_types: VisitTransitionSet,
+) -> Result<HistoryVisitInfos> {
+    let allowed_types = exclude_types.complement();
+    let infos = db.query_rows_and_then_named_cached(
+        "SELECT h.url, h.title, v.visit_date, v.visit_type
+         FROM moz_places h
+         JOIN moz_historyvisits v
+           ON h.id = v.place_id
+         WHERE ((1 << v.visit_type) & :allowed_types) != 0
+         ORDER BY v.visit_date DESC
```

Comment:
Is this guaranteed to be a consistent strict ordering, or would it be possible for different invocations of this query to return results in a different order if they happen to have the same `visit_date`?  It may be worth ordering by primary key here as well to enforce a strict ordering.